### DISCUSSION TITLE
出现刷任务点之后停止工作（automaton.utils:dfs:78 - [刷任务点]）是什么问题呢？

### DISCUSSION THREAD

**Discussion by Xiangyuan-Xie** [Mufanc/iSmartAuto#20] [Q&A]
翻了翻发现作者在readme的上一版写过父节点不存在这个问题不会影响程序进行，但是我在运行中出现下面内容之后就停止了（处于待输入命令状态），于是乎也就没有刷到任务点……

> 2022-05-16 23:59:52.391 | INFO     | automaton.spider.spider:get_tasks:118 - [构建任务树] | 正在构建任务树...
> 2022-05-16 23:59:52.393 | WARNING  | automaton.spider.spider:get_tasks:126 - [构建任务树] | Unit 1 Knowledge is Great[id:1606e9ea2469426fa98ad6d290c91e3b] 父节点不存在
> 2022-05-16 23:59:52.394 | WARNING  | automaton.spider.spider:get_tasks:126 - [构建任务树] | Unit 2 Employment[id:5ebfb00164b349a786f9741fd8e3f632] 父节点不存在
> 2022-05-16 23:59:52.396 | WARNING  | automaton.spider.spider:get_tasks:126 - [构建任务树] | Unit 3 Advertising[id:b50e471179924780a748b0c3b3862f48] 父节点不存在
> 2022-05-16 23:59:52.397 | WARNING  | automaton.spider.spider:get_tasks:126 - [构建任务树] | Unit 4 Catastrophe[id:f67f36d7c2c44e4bbb8344c3c0018c07] 父节点不存在
> 2022-05-16 23:59:52.399 | WARNING  | automaton.spider.spider:get_tasks:126 - [构建任务树] | Unit 5 Ways to Success[id:e0e9b30713404cbca3cb75688d4107a6] 父节点不存在
> 2022-05-16 23:59:52.400 | WARNING  | automaton.spider.spider:get_tasks:126 - [构建任务树] | Unit 6 All on the Mind[id:9c3ce2acefb84e48b4b6b5fffe4467bc] 父节点不存在
> 2022-05-16 23:59:52.401 | WARNING  | automaton.spider.spider:get_tasks:126 - [构建任务树] | Unit 7  Finding Happiness[id:fec67009b7c445e2aafd5bab160e6911] 父节点不存在
> 2022-05-16 23:59:52.403 | WARNING  | automaton.spider.spider:get_tasks:126 - [构建任务树] | Unit 8 Are We Safe on the Internet?[id:dc839ebaefdf4ea381b95661550c2422] 父节点不存在
> 2022-05-16 23:59:52.404 | SUCCESS  | automaton.spider.spider:get_tasks:130 - [构建任务树] | 构建任务树完成
> 2022-05-16 23:59:52.405 | SUCCESS  | automaton.spider.spider:get_tasks:131 - [获取任务列表] | 获取任务列表完成
> 2022-05-16 23:59:52.406 | INFO     | automaton.utils:dfs:78 - [刷任务点] |  大学体验英语视听说2

是在刷任务点的时候出现的问题嘛，以下是我找到的刷任务点的函数，应该怎么去修改呢？
Python新手，感谢解惑！（总算排版排得让人能看了

```
async def _flash(course_id, book_id, spider):
    try:
        async def dfs(node, depth=0):
            task = node.task
            logger.info(f"[刷任务点] | {PLACEHOLDER * depth} {task['name']}")
            if _paper_config['skip-finished'] and task['unitStudyPercent'] == 100:
                logger.info(f"[刷任务点] | {PLACEHOLDER * depth} 跳过")
                return
            if 'paperId' in task:  # 如果有任务则提交
                chapter_id = task['chapterId']
                task_id = task['id']
                score, time = await _random(spider, task['paperId'])
                result = await page.submit(book_id, chapter_id, task_id, score, time, 100, user_id)
                if result['wasThrown'] or not result['result']['value']:
                    logger.warning(f'[刷任务点] | 任务[{task["name"]}]可能提交失败，请留意最终结果')
            for ch in node.child:
                await dfs(ch, depth + 1)

        browser = await Browser.connect()
        page = await browser.wait_for_page(r'https?://pc\.ismartin\.com.*')

        # With Spider
        await spider.login(**configs['user'])
        user_id = (await spider.user_info())['data']['uid']
        book_type = (await spider.book_info(book_id))['bookType']
        root = await spider.get_tasks(book_id, book_type, course_id)
        await dfs(root)
    except Exception:
        logger.warning(f'[刷任务点] | 刷任务点出错：\n{traceback.format_exc()}')
```

**Comment by pianai520** [comment]
同问 想问下题主解决了吗

**Reply by Xiangyuan-Xie** [reply]
并没有……
我的猜测是可能在构建的任务树并没有内容，所以到了刷任务点阶段就终止了，可能跟与教材差异有关，但是我没想出来要怎么改……

**Reply by CNorato** [reply]
我用的也是这个系列的教材, 这系列的教材大部分都是需要读的题目, 然后简单调试了一下发现在task中没有paperId这个属性, 因此直接就跳过了...
另外我将构建任务树那里改为了这样, 对这个系列的教材构建出来节点会更好看
```python
for task_id in id_record:
    node = id_record[task_id]
    # node_name = f'{node.task.get("name","")}[id:{node.task["id"]}]'
    if 'parent_id' in node.task:
        if node.task['parent_id'] in id_record:
            id_record[node.task['parent_id']].child.append(node)
        else:
            root.child.append(node)
    else:
        root.child.append(node)
````

**Reply by CNorato** [reply]
@WhiteCloudAB 我把tasks截取出来了, 不知道有没有用
[tasks.zip](https://github.com/Mufanc/iSmartAuto2/files/8767697/tasks.zip)

**Reply by yinzeidog** [reply]
我也出现这种问题了 请问有解决吗

**Reply by fangsancs** [reply]
作者的代码应该是是根据每道题的“paperId”属性到网页上(（[https://xot-api.ismartlearning.cn/client/textbook/paperinfo）)寻找
答案，而“booktype1”的书的题目没有“paperId”这个属性，导致以下代码没有执行：（这一串代码的”if 'paperId' in task“结果为假
，代码没有执行。）
if 'paperId' in task: # 如果有任务则提交
chapter_id = task['chapterId']
task_id = task['id']
score, time = await _random(spider, task['paperId'])
result = await page.submit(book_id, chapter_id, task_id, score, time, 100, user_id)
if result['wasThrown'] or not result['result']['value']:
logger.warning(f'[刷任务点] | 任务[{task["name"]}]可能提交失败，请留意最终结果')

基于以上猜想，我尝试使用”booktype1“的书的题目的”id“属性来作为判断条件，虽然”if 'paperId' in task“结果为真，但这样以后“
获取任务点出错”，可能ismart的题目的答案是基于“paperId”属性一 一对应的。但是我猜想，”booktype1“的书可能是根据题目的
”id“属性来一 一对应答案，我猜想现在的代码需要修改题目的答案所在的网址，脚本应该就能正常运行。但是，我不会抓包，也
不懂得如何获得答案所在的网址，在座各位有能助我一臂之力的高手吗？（作者大大怎么获得答案的网址的，可以教教吗；能区
分不同题目的也只有“id”属性了，但是用这个属性去作者给的答案网址去找却会报错，所以答案网址不止一个是我的猜想；我也
研究了很久，最后一关是我的盲区了QAQ）

**Comment by appleDGY233** [comment]
### **BookType为1 的这一类的教材都有问题  **
起初我与题主遇到了同样的问题，“父节点不存在”且无任何任务点刷的问题，运行信息与题主相同。
后来翻看作者标出的 [父节点不存在] 解决方法
修改完之后的运行结果为

> 2022-05-30 02:05:21.997 | SUCCESS  | automaton.spider.spider:book_info:214 - [获取书籍信息] | 获取书籍信息完成
2022-05-30 02:05:21.998 | INFO     | automaton.spider.spider:get_tasks:118 - [构建任务树] | 正在构建任务树...
2022-05-30 02:05:22.097 | SUCCESS  | automaton.spider.spider:get_tasks:140 - [构建任务树] | 构建任务树完成
2022-05-30 02:05:22.097 | SUCCESS  | automaton.spider.spider:get_tasks:141 - [获取任务列表] | 获取任务列表完成
2022-05-30 02:05:22.098 | INFO     | automaton.utils:dfs:78 - [刷任务点] |  大学体验英语视听说2
2022-05-30 02:05:22.099 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .    Unit 1 Knowledge is Great
2022-05-30 02:05:22.100 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .    Lead-in
2022-05-30 02:05:22.101 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .    Lead-in
......中间省略一大段相同的......
2022-05-30 02:05:22.165 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .    Video Studio
2022-05-30 02:05:22.165 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .    Unit 6 All on the Mind

虽然任务点都识别出来了，可还是没有刷上

以下这个可以正常刷的教材（BookType为0的教材）运行结果作为对比
> 2022-05-30 01:36:26.885 | SUCCESS  | automaton.spider.spider:get_tasks:130 - [构建任务树] | 构建任务树完成
2022-05-30 01:36:26.885 | SUCCESS  | automaton.spider.spider:get_tasks:131 - [获取任务列表] | 获取任务列表完成
2022-05-30 01:36:26.886 | INFO     | automaton.utils:dfs:78 - [刷任务点] |  大学体验英语（第四版）听说教程2
2022-05-30 01:36:26.887 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .    Unit 1 Protectors
2022-05-30 01:36:26.888 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .   .    WARM UP
2022-05-30 01:36:26.889 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .   .   .    LEAD-IN
**2022-05-30 01:36:26.890 | INFO     | automaton.spider.spider:get_paper:138 - [获取任务点] | 正在获取任务点信息...
2022-05-30 01:36:26.946 | DEBUG    | automaton.spider.spider:get_paper:145 - [获取任务点] | ST-24352862-QdLlVle0CzHb93iLE6WK-cas01.example.org**
......中间省略一大段相同的......
2022-05-30 01:36:36.313 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .   .   .    IMITATING
2022-05-30 01:36:36.314 | INFO     | automaton.spider.spider:get_paper:138 - [获取任务点] | 正在获取任务点信息...
2022-05-30 01:36:36.370 | DEBUG    | automaton.spider.spider:get_paper:145 - [获取任务点] | ST-3462756-VR0BDEsRWl67TkFgO0DF-cas01.example.org
2022-05-30 01:36:36.427 | SUCCESS  | automaton.spider.spider:get_paper:162 - [获取任务点] | 获取任务点信息完成
2022-05-30 01:36:36.470 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .   .    6E Extended reading
2022-05-30 01:36:36.471 | INFO     | automaton.utils:dfs:78 - [刷任务点] | .   .   .    READING COMPREHENSION
2022-05-30 01:36:36.472 | INFO     | automaton.spider.spider:get_paper:138 - [获取任务点] | 正在获取任务点信息...
2022-05-30 01:36:36.530 | DEBUG    | automaton.spider.spider:get_paper:145 - [获取任务点] | ST-3462757-xlU16M44RNx3ZKwMNdfl-cas01.example.org
2022-05-30 01:36:36.588 | SUCCESS  | automaton.spider.spider:get_paper:162 - [获取任务点] | 获取任务点信息完成

可以发现并没有
> 2022-05-30 01:36:36.314 | INFO     | automaton.spider.spider:get_paper:138 - [获取任务点] | 正在获取任务点信息...
2022-05-30 01:36:36.370 | DEBUG    | automaton.spider.spider:get_paper:145 - [获取任务点] | ST-3462756-VR0BDEsRWl67TkFgO0DF-cas01.example.org

这一类的运行结果出现.。
检查utils.py中此段代码发现
```

async def _flash(course_id, book_id, spider):
    try:
        async def dfs(node, depth=0):
            task = node.task
            logger.info(f"[刷任务点] | {PLACEHOLDER * depth} {task['name']}")
            if _paper_config['skip-finished'] and task['unitStudyPercent'] == 100:
                logger.info(f"[刷任务点] | {PLACEHOLDER * depth} 跳过")
                return
            logger.success("success2")
            if 'paperId' in task:  # 如果有任务则提交
                chapter_id = task['chapterId']
                task_id = task['id']
                score, time = await _random(spider, task['paperId'])
                result = await page.submit(book_id, chapter_id, task_id, score, time, 100, user_id)
                if result['wasThrown'] or not result['result']['value']:
                    logger.warning(f'[刷任务点] | 任务[{task["name"]}]可能提交失败，请留意最终结果')
            for ch in node.child:
                await dfs(ch, depth + 1)

```
" if 'paperId' in task"被判定为假，并未执行
也就是说脚本只是把目录遍历了一遍并没有上传分数
检查“spider.py”后发现
```
 async def get_paper(self, paper_id):  # 获取任务点信息（包括题目和答案）
        try:
            logger.info('[获取任务点] | 正在获取任务点信息...')
            ticket = (await self.post(
                'http://sso.ismartlearning.cn/v1/serviceTicket',
                data={
                    'service': 'http://xot-api.ismartlearning.cn/client/textbook/paperinfo'
                }
            )).json()['data']['serverTicket']
            logger.debug(f'[获取任务点] | {ticket}')
            paper_info = (await self.post(
                'http://xot-api.ismartlearning.cn/client/textbook/paperinfo',
                params={
                    'ticket': ticket
                },
                data={
                    'paperId': paper_id
                },
                headers={
                    'Origin': 'http://me.ismartlearning.cn',
                    'Referer': 'http://me.ismartlearning.cn/',
                    'X-Requested-With': 'XMLHttpRequest',
                    'Accept-Encoding': 'gzip, deflate'
                }
            )).json()['data']
            # logger.debug(f'[获取任务点] | {json.dumps(paper_info, indent=4)}')
            logger.success('[获取任务点] | 获取任务点信息完成')
            return paper_info
        except Exception:
            logger.warning(f'[获取任务点] | 获取任务点出错：\n{traceback.format_exc()}')

```
这段代码并未执行（或是执行失败）
到这里我就不会修了
### 求帮助

**Reply by 5gentle5** [reply]
那个获取答案信息的网页我这边打开的是空白的：http://sso.ismartlearning.cn/v1/serviceTicket
会不会是根本没有获取到答案，所有没有执行？

**Reply by 2977791420** [reply]
你现在搞好了吗，大佬

**Comment by pianai520** [comment]
请问解决了嘛

**Comment by fangsancs** [comment]
作者的代码应该是是根据每道题的“paperId”属性到网页上(（[https://xot-api.ismartlearning.cn/client/textbook/paperinfo）)寻找
答案，而“booktype1”的书的题目没有“paperId”这个属性，导致以下代码没有执行：（这一串代码的”if 'paperId' in task“结果为假
，代码没有执行。）
if 'paperId' in task: # 如果有任务则提交
chapter_id = task['chapterId']
task_id = task['id']
score, time = await _random(spider, task['paperId'])
result = await page.submit(book_id, chapter_id, task_id, score, time, 100, user_id)
if result['wasThrown'] or not result['result']['value']:
logger.warning(f'[刷任务点] | 任务[{task["name"]}]可能提交失败，请留意最终结果')

基于以上猜想，我尝试使用”booktype1“的书的题目的”id“属性来作为判断条件，虽然”if 'paperId' in task“结果为真，但这样以后“
获取任务点出错”，可能ismart的题目的答案是基于“paperId”属性一 一对应的。但是我猜想，”booktype1“的书可能是根据题目的
”id“属性来一 一对应答案，我猜想现在的代码需要修改题目的答案所在的网址，脚本应该就能正常运行。但是，我不会抓包，也
不懂得如何获得答案所在的网址，在座各位有能助我一臂之力的高手吗？（作者大大怎么获得答案的网址的，可以教教吗；能区
分不同题目的也只有“id”属性了，但是用这个属性去作者给的答案网址去找却会报错，所以答案网址不止一个是我的猜想；我也
研究了很久，最后一关是我的盲区了QAQ）

**Reply by Mufanc** [reply]
已上传实现细节：[https://blog.mufanc.xyz/posts/3824/](https://blog.mufanc.xyz/posts/3824/)

**Reply by TWLW9784** [reply]
> e
弄好了嘛
